packages/runtime/src/runtimes/react/react-runtime.ts

```diff
@@ -19,7 +20,41 @@ function validateComponentType(type: string, component?: ComponentType): void {
   }
 }
 
+type ComponentRegistrationMeta<
+  P extends Record<string, LegacyDescriptor | UnifiedControlDefinition>,
+> = {
+  type: string
+  label: string
+  icon: ComponentIcon
+  hidden: boolean
+  props?: P
+}
+
+export type ComponentRegistration<
+  P extends Record<string, LegacyDescriptor | UnifiedControlDefinition>,
+> = {
+  component: ComponentType<{ [K in keyof P]: DescriptorValueType<P[K]> }>
+  meta: ComponentRegistrationMeta<P>
+}
+
 export class ReactRuntime extends RuntimeCore {
+  static connect<
```

Comment:
The more I think about this change, the more I question whether it makes sense to ship it incrementally, or whether it would be better folded into our broader effort to remake the runtime and decouple it from Next.js.

While this doesn’t significantly *increase* the complexity of the component registration model, it also doesn’t reduce the number of concepts and entities users need to learn or define to get started with Makeswift. At the same time, it introduces another round of API changes for existing users, without offering a clear, immediate benefit.

If we weren’t already planning a larger runtime overhaul in the near future, I’d feel more comfortable polishing and shipping this as a standalone improvement. But given that we are, I think it’s worth raising the question.

Reply:
> I feel like we're holding back on something that we know is clearly better

Yeah, I don't feel that way, which is why I wrote what I wrote. I tried to polish this into a shippable state and imagine trying to justify this change to the users, and I couldn't sell it to myself. It might be just me, but that's where my original comment was coming from.

Reply:
That's alright. I'm ok with us closing this and bringing it back later when it makes more sense, then. I'll delegate to you guys on this.